doc/HowToPipeline.md

```diff
@@ -0,0 +1,2040 @@
+                         pipelineConnections=ApertureTaskConnections):
+    ...
+    doLocalBackground = pexConfig.Field(doc="Should the background be added before doing photometry",
+                                        dtype=bool, default=False)
+```
+
+The `\_\_init\_\_` method of the connection class is given an instance of the tasks config class after all
+overrides have been applied. This provides us an opportunity change the behavior of the connection class
+according to various config options.
+
+```python
+class ApertureTaskConnections(pipeBase.PipelineTaskConnections,
+                              dimensions=("visit", "detector", "abstract_filter", "skymap")):
+    ...
+    background = pipeBase.connectionTypes.Input(doc="Background model for the exposure",
+                                                storageClass="Background",
+                                                name="calexpBackground",
+                                                dimensions=("visit", "detector", "abstract_filter", "skymap"))
+    ...
+
+    def __init__(self, *, config=None):
+        super().__init__(config=config)
+        
+        if config.doLocalBackground is False:
+            self.inputs.remove("background")
+```
+
+Our connection class now looks at the value of `doLocalBackground` on the config object and if it is false,
+removes it from the connection instances list of input connections. Connection classes keep track of what
+connections are defined in sets. Each set contains the variable names of a connection, and the sets
+themselves are identified by the type of connection they contain. In our example we are modifying the set
+of input connections. The names for each of the sets are as follows:
+
+* initInputs
+* initOutputs
+* inputs
+* prerequisiteInputs
+* outputs 
+
+The last step in modifying our task will be to update the run method to take into account that a background
+may or may not be supplied.
+
+```python
+...
+import typing
+
+import lsst.afw.math as afwMath
+
+...
+
+class ApertureTask(pipeBase.PipelineTask):
+    
+    ...
+
+    def run(self, exposure: afwImage.Exposure, inputCatalog: afwTable.SourceCatalog,
+            background: typing.Union[None, afwMath.BackgroundList] = None) -> pipeBase.Struct:
+        # If a background is supplied, add it back to the image so local background subtraction
+        # can be done.
+        if background is not none:
+            exposure.image.array += background.getImage()
+
+        ...
+
+        # Loop over each record in the catalog
+        for source in outputCatalog:
+
+            ...
+
+            # Do local background subtraction
+            if background is not None:
+                outerAn = distance < 2.5*self.apRad
+                innerAn = distance < 1.5*self.apRad
+                annulus = outerAn - innerAn
+                localBackground = np.mean(exposure.image.array*annulus)
+                flux -= np.sum(mask)*localBackground
```

Comment:
It didn't occur to me until now, but you might want to just hide all of these irrelevant algorithmic details behind verbosely-named free functions (e.g. `doLocalBackgroundSubtraction(exposure)`) that you don't try to spell out in the examples (you could, at least sometimes, explicitly import them from dummy modules instead).